charms/kserve-controller/src/charm.py

```diff
@@ -0,0 +1,311 @@
+        """Define and start a workload using the Pebble API.
+
+        Learn more about Pebble layers at https://github.com/canonical/pebble
+        """
+        try:
+            self.gen_certs(self.model.name, self.app.name)
+
+            self.controller_container.push("/tmp/k8s-webhook-server/serving-certs/tls.crt", Path("/run/cert.pem").read_text(), make_dirs=True)
+            self.controller_container.push("/tmp/k8s-webhook-server/serving-certs/tls.key", Path("/run/server.key").read_text(), make_dirs=True)
+
+            update_layer(
+                self._controller_container_name,
+                self.controller_container,
+                self._controller_pebble_layer,
+                self.log,
+            )
+        except ErrorWithStatus as e:
+            self.model.unit.status = e.status
+            if isinstance(e.status, BlockedStatus):
+                self.log.error(str(e.msg))
+            else:
+                self.log.info(str(e.msg))
+
+        # TODO determine status checking if rbac proxy is also up
+        self.unit.status = ActiveStatus()
+
+    def _on_kube_rbac_proxy_ready(self, event):
+        """Define and start a workload using the Pebble API.
+
+        Learn more about Pebble layers at https://github.com/canonical/pebble
+        """
+        try:
+            update_layer(
+                self._rbac_proxy_container_name,
+                self.rbac_proxy_container,
+                self._rbac_proxy_pebble_layer,
+                self.log,
+            )
+        except ErrorWithStatus as e:
+            self.model.unit.status = e.status
+            if isinstance(e.status, BlockedStatus):
+                self.log.error(str(e.msg))
+            else:
+                self.log.info(str(e.msg))
+
+
+        #TODO determine status checking if controller is also up
+        self.unit.status = ActiveStatus()
+
+    def _on_install(self, event):
+        try:
+            self.unit.status = MaintenanceStatus("Creating k8s resources")
+            self.k8s_resource_handler.apply()
+        except ApiError as e:
+            self.log.error(e)
+        self.model.unit.status = ActiveStatus()
+
+    def _on_remove(self, _):
+        self.unit.status = MaintenanceStatus("Removing k8s resources")
+        k8s_resources_manifests = self.k8s_resource_handler.render_manifests()
+        try:
+            delete_many(
+                    self.k8s_resource_handler.lightkube_client,
+                    k8s_resources_manifests,
+                )
+        except ApiError as e:
+            self.log.warning(f"Failed to delete resources, with error: {e}")
+            raise e
+        self.unit.status = MaintenanceStatus("K8s resources removed")
+
+    def gen_certs(self, namespace, service_name):
+        if Path("/run/cert.pem").exists():
+            self.log.info("Found existing cert.pem, not generating new cert.")
+            return
+    
+        Path("/run/ssl.conf").write_text(
```

Comment:
Just a thought to keep code cleaner: create template file for `ssl.conf` and load it instead of putting it in the code. Similar we do with other templates I did it for Seldon certs.
Just cosmetics, can wait till final PR.